### ISSUE TITLE
Multidimensional Arrays for Unit Testing

### ISSUE THREAD

**Issue by rosenbrockc** [rosenbrockc/fortpy#41]
Currently, fortpy only supports multi-dimensional arrays through the ragged array options. This forces the user to split their 3+ dimension array into a set of files, each of which is 2D. For example, a matrix of point group operations with size `(48, 3, 3)` would be split into 48 files, each of which has a `3 x 3` matrix in it.

The problem of handling multi-D arrays in files is not new. Here are some of the common solutions:

1. Write the shape of the array on the first line in the file and then write 2D slices of the array, in order, to the file. The shape allows quick reconstruction using a function like `numpy.reshape`.
2. Use the HDF5 file standard to binary serialize the resulting matrices completely. There are strong libraries supporting python and fortran for HDF5; also the resulting files would be dense and occupy less space. _Disadvantage_: the files are no longer human-readable. This means that an HDF5 file generating as "model" output can't easily be verified without browsing the HDF5 data set using some other tool to make sure the results are correct.
3. Use NetCDF, which is similar to the idea of HDF5 (compression, binary serialization, etc.); its advantage is that it was particularly designed to handle arrays of varying dimension.

In some respects, using HDF5 seems like a really good idea for all array-valued outputs/inputs. A unit test could have all its outputs etc. saved into a single HDF5 data set file. We would have to update fortpy to be able to handle the comparisons well. Since HDF provides a python interface to the file contents, this upgrade would be tiny. Most of the work would be changing the way that data gets saved. We could do this by just editing the `fortpy.f90` file's `pysave` interface. The unit tests would also use up less disk space then. We would also need to find a good HDF file browser helping developers check their model output.

Thoughts?

**Comment by wsmorgan** [comment]
Getting the input and outputs into a single file would be very nice except for the fact that if we went that way the only way we would be able to build input files for our unit tests would be to use fortpy. This would be inconvenient when we use things like mathematica to develop unit tests (which I have been doing for some of subroutines I've been testing). So even if it's less convenient I think the more generally accessible option is to use the first option and save/read the array's in slices. 

Another possible option would be to have fortpy support either format, HDF5 or slices and have the user add a modifier to the input/output tag that would tell it which of the two it needed to use.

**Comment by rosenbrockc** [comment]
For the short term, we are going to just handle the arrays in plain text and leave the HDF5 for now. We will reserve the double pound symbol `##` as the indicator for array shaping information. This is how it will work: 

```
## 3 2 2
## 1,:,:
0 1
1 0
## 2,:,:
-1 1
1 -1
## 3,:,:
2 3
4 5
```

- The first line specifies the total shape of the array that can be used for allocating it in memory. It is just a list of the dimensions in the order that they would be entered in Fortran directly.
- The multi-dimensional array is specified in blocks of 2D arrays. Each 2D array block should be preceded by the index specifying where its data should be copied. Fortpy will splice the exact form of the index into the variable name on the left of the assignment. `variable(3,:,:) = read(values)`.

This file format should be able to handle arrays of any dimension. In order to implement this, we will need:

1. Additions to the interfaces `fpy_read`, `fpy_read_p` and `fpy_read_f` that accept variables of dimensions 3 through 7 (which is the maximum supported dimensionality in fortran).
2. Additions to the `fpy_save` interface for dimensions 3 through 7 that writes out in this format.
3. A new file template for `integer.xml` and `float.xml` that handle the multi-D format.

At this point, it may be worthwhile to write a python module that can create the `fortpy.f90` file without making errors. Then we will have uniformity across the implementation when the only differences are adding/removing `allocatable` or `pointer` modifiers. This will also allow the problem with `logical` not being supported to be addressed easily as well as `complex` types in the future, where the values have to be separated in real and imaginary parts.

**Comment by rosenbrockc** [comment]
Support available from Revision 1.6.0
